### DISCUSSION TITLE
services 'OPTIONS' request

### DISCUSSION THREAD

**Discussion by farfromrefug** [valhalla/valhalla#4955] [Q&A] [answered]
The valhalla  app  https://github.com/nilsnolde/valhalla-app makes `OPTIONS` requests for `height` and `locate`. It works fine on https://valhalla.openstreetmap.de
However when i try to run the valhalla app with my own vahalla server i get errors for `OPTIONS` requests : 
```
{
	"error_code": 101,
	"error": "Try a POST or GET request instead",
	"status_code": 405,
	"status": "Method Not Allowed"
}
```
Does anyone knows how to make it work ? It could be `Access-Control-Allow-Methods`  but i dont see how to enforce it

**Comment by jthiard** [comment]
The valhalla web service does not support any CORS headers / preflight request by itself (see #3951).
You will need a reverse proxy in front of the valhalla service. As an example, using nginx to support preflight requests from all origins:

```nginx
server {
    ...
    location / {
        if ($request_method = 'OPTIONS') {
            # Preflight request : directly respond with nginx
            add_header 'Access-Control-Allow-Methods' 'HEAD, GET, POST, OPTIONS';
            add_header 'Access-Control-Allow-Origin' '*';
            add_header 'Content-Length' 0;
            return 204;
        }
        # Normal request : pass to valhalla
        proxy_pass http://localhost:8080/;
    }
}
```

**Reply by kevinkreiser** [reply]
yes this is what most people do, the main reason this makes sense is because most people who do this in production will need:

1. rate limiting
2. authentication

the built in http server does not have that and any one that does have that probably supports the cors preflight. long term itd be nice to support but.. probably not something happening any time soon

**Comment by chrstnbwnkl** [comment, marked as answer]
Linking https://github.com/kevinkreiser/prime_server/pull/118 for completeness
